File: maven.js
class Maven {
	constructor(major, minor = undefined, patch = undefined, extra = undefined, tagln = undefined) {
		if (minor !== undefined) {
			this.major = major
			this.minor = minor
			this.patch = patch
			this.extra = extra
			this.tagln = tagln
		} else if (typeof major == "string") {
			let objs = /^(\d+?)\.(\d+?)(?:\.(\d+?))?(?:-(.+?)(?: |$))?(?: ?"(.*?)")?$/.exec(major)
			if (objs === null) { throw SyntaxError("Malformed Maven version") }
			this.major = objs[1]
			this.minor = objs[2]
			this.patch = objs[3]
			this.extra = objs[4]
			this.tagln = objs[5]
		} else if (major instanceof Maven) {
			this.major = major.major
			this.minor = major.minor
			this.patch = major.patch
			this.extra = major.extra
			this.tagln = major.tagln
		} else {
			throw new SyntaxError("Invalid constructor input")
		}
	}
	getId() {
		return this.major + "." + this.minor + this.patch ? "." + this.patch : "" + this.extra ? "-" + this.extra : ""
	}
	getName() {
		return this.getVersionId() + this.tagln ? ' "' + this.tagln + '"' : ""
	}
	gt(other) {
		other = new Maven(other)
		if (this.major != other.major) { return this.major > other.major }
		if (this.minor != other.minor) { return this.minor > other.minor }
		if (! this.patch && !! other.patch) { return true }
		if (! other.patch && !! this.patch) { return false }
		if (this.patch != other.patch) { return this.patch > other.patch }
		if (! this.extra && !! other.extra) { return true }
		if (! other.extra && !! this.extra) { return false }
		if (this.extra != other.extra) { return this.extra > other.extra }
		return false
	}
	eq(other) {
		other = new Maven(other)
		return this.major == other.major && this.minor == other.minor && this.patch == other.patch && this.extra == other.extra && this.tagln == other.tagln
	}
	gte(other) { return this.eq(other) || this.gt(other) }
	neq(other) { return !this.eq(other) }
	lte(other) { return !this.gt(other) }
	lt(other) { return !this.gte(other) }
}

File: maven.py
import re

class Maven:
    def __init__(self, maj, /, min = None, pat = None, ext = None, *, tag = None):
        if type(min) != "NoneType":
            self.major = maj
            self.minor = min
            self.patch = pat
            self.extra = ext
            self.tagln = tag
        elif type(maj) == "Maven":
            self.major = maj.major
            self.minor = maj.minor
            self.patch = maj.patch
            self.extra = maj.extra
            self.tagln = maj.tagln
        elif type(maj) == "str":
            objs = re.search('^(\d+?)\.(\d+?)(?:\.(\d+?))?(?:-(.+?)(?: |$))?(?: ?"(.*?)")?$', maj)
            if type(objs) == "NoneType":
                raise SyntaxError("Malformed Maven version")
            self.major = objs.groups()[0]
            self.minor = objs.groups()[1]
            self.patch = objs.groups()[2]
            self.extra = objs.groups()[3]
            self.tagln = objs.groups()[4]
        else:
            raise SyntaxError("Invalid constructor input")
    def get_id(self):
        return self.major + "." + self.minor + ("." + self.patch if self.patch else "") + ("-" + self.extra if self.extra else "")
    def get_name(self):
        return self.get_id() + (' "' + self.tagln + '"' if self.tagln else "")
    def __gt__(self, other):
        other = Maven(other)
        if self.major != other.major: return self.major > other.major
        if self.minor != other.minor: return self.minor > other.minor
        if self.patch and not other.patch: return True
        if other.patch and not self.patch: return False
        if self.patch != other.patch: return self.patch > other.patch
        if self.extra and not other.extra: return True
        if other.extra and not self.extra: return False
        for a, b in zip(self.extra, other.extra):
            if a != b: return a > b
        return False
    def __eq__(self, other):
        other = Maven(other)
        return self.major == other.major and self.minor == other.minor and self.patch == other.patch and self.extra == other.extra
    def __le__(self, other): return not self > other
    def __ne__(self, other): return not self == other
    def __ge__(self, other): return self > other or self == other
    def __lt__(self, other): return not self >= other